cpu/esp_common/periph/uart.c

```diff
@@ -166,30 +168,34 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
     assert(uart < UART_NUMOF);
 
 #ifndef MCU_ESP8266
-    /* reset the pins when they were already used as UART pins */
-    if (gpio_get_pin_usage(uart_config[uart].txd) == _UART) {
-        gpio_set_pin_usage(uart_config[uart].txd, _GPIO);
-    }
-    if (gpio_get_pin_usage(uart_config[uart].rxd) == _UART) {
-        gpio_set_pin_usage(uart_config[uart].rxd, _GPIO);
-    }
-
-    /* try to initialize the pins as GPIOs first */
-    if ((uart_config[uart].txd != GPIO_UNDEF &&
-         gpio_init(uart_config[uart].txd, GPIO_OUT)) ||
-        (uart_config[uart].rxd != GPIO_UNDEF &&
-         gpio_init(uart_config[uart].rxd, GPIO_IN_PU))) {
-        return -1;
-    }
+    assert(uart_config[uart].txd != GPIO_UNDEF);
+    assert(uart_config[uart].rxd != GPIO_UNDEF);
```

Comment:
There are use cases of TX only / RX only UARTs.
This is generally allowed by other UART implementations, e.g sam0_common.

Reply:
OK understand, also the ESP8266 has a second UART device which can only be TX.

But this change is not a regression from the previous implementation, which didn't really work for TX-only and RX-only interface, because it didn't handle the TX and RX pin separately.

For now, this PR would fix the problem in the release version. The handling of TX- or RX-only UARTs could be improved as a follow-up.

Reply:
Ok, let’s do that as a follow-up then

bors merge